cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -94,18 +96,28 @@ func PerformPostUpgradeTasks(client clientset.Interface, cfg *kubeadmapi.InitCon
 		errs = append(errs, err)
 	}
 
-	// Upgrade kube-dns/CoreDNS and kube-proxy
-	if err := dns.EnsureDNSAddon(&cfg.ClusterConfiguration, client); err != nil {
-		errs = append(errs, err)
-	}
-	// Remove the old DNS deployment if a new DNS service is now used (kube-dns to CoreDNS or vice versa)
-	if err := removeOldDNSDeploymentIfAnotherDNSIsUsed(&cfg.ClusterConfiguration, client, dryRun); err != nil {
-		errs = append(errs, err)
+	if !features.Enabled(cfg.ClusterConfiguration.FeatureGates, features.AddonInstaller) {
+		// Upgrade kube-dns/CoreDNS and kube-proxy
+		if err := dns.EnsureDNSAddon(&cfg.ClusterConfiguration, client); err != nil {
+			errs = append(errs, err)
+		}
+		// Remove the old DNS deployment if a new DNS service is now used (kube-dns to CoreDNS or vice versa)
+		if err := removeOldDNSDeploymentIfAnotherDNSIsUsed(&cfg.ClusterConfiguration, client, dryRun); err != nil {
+			errs = append(errs, err)
+		}
+
+		if err := proxy.EnsureProxyAddon(&cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, client); err != nil {
+			errs = append(errs, err)
+		}
 	}
 
-	if err := proxy.EnsureProxyAddon(&cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, client); err != nil {
-		errs = append(errs, err)
+	if features.Enabled(cfg.ClusterConfiguration.FeatureGates, features.AddonInstaller) {
```

Comment:
Is it better to use `else` clause instead of another `if` clause? Because this condition is just opposite previous one (L99).